### PR TITLE
ratelimit support query parameter match

### PR DESCRIPTION
**Pull request by StarryVae**
Signed-off-by: wangkai19 <wangkai19@corp.netease.com>

Commit Message: ratelimit support query parameter match
Additional Description: N/A
Risk Level: low
Testing: ut
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by StarryVae** [comment]
issue #23302 ratelimit supports query parameter match, here is the api refer to `HeaderValueMatch`, please review the api first, thanks. cc @wbpcode @mattklein123

**Comment by StarryVae** [comment]
ping @markdroth

**Comment by StarryVae** [comment]
ping @markdroth

**Comment by wbpcode** [comment]
Hi, @markdroth could you take a look for this new proto API when you have free time? Thanks.

**Comment by KBaichoo** [comment]
friendly ping @markdroth , it seems like this one is waiting on your input

**Comment by phlax** [comment]
@markdroth @envoyproxy/api-shepherds i think this one is waiting on api-shep approval

/wait-any

**Comment by StarryVae** [comment]
friendly ping @markdroth , cc @phlax @mattklein123 , thanks.

**Comment by wbpcode** [comment]
Seems that @markdroth is too busy and re-assign this PR.

**Comment by jmarantz** [comment]
@adisuissa @wbpcode can you take a look?

File: api/envoy/config/route/v3/route_components.proto
Comment (adisuissa) [inline api/envoy/config/route/v3/route_components.proto:2102]: A high-level question before going into this. Why isn't this a `rate_limit_descriptor` extension?
Making it an extension will avoid this from being linked if not needed.

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this.
Left a high-level question, but other than that the `QueryParameterValueMatch` API LGTM.

File: api/envoy/config/route/v3/route_components.proto
Comment (StarryVae) [reply api/envoy/config/route/v3/route_components.proto:2102]: @adisuissa thanks! but i am not familiar with `rate_limit_descriptor` extension, can it also support various types of matches(prefix, suffix, regex...) like QueryParameterValueMatch?

Comment (adisuissa) [reply api/envoy/config/route/v3/route_components.proto:2102]: There's an extension [here](https://github.com/envoyproxy/envoy/blob/a1606b267c27a6ffc057ac3abe7ae38734088265/source/extensions/rate_limit_descriptors/expr/config.h) that introduces a CEL rate-limiter.

That said, if you think there's a good reason for this to be integral part of Envoy's core, LMK.

Comment (StarryVae) [reply api/envoy/config/route/v3/route_components.proto:2102]: i think `rate_limit_descriptor` extension does not have all the capabilities of `QueryParameterMatcher` which supports different types of match for each query parameter.

**Comment by alyssawilk** [comment]
@adisuissa ping!

**Review by adisuissa — COMMENTED** [review]
Apologies for the late reply.
/lgtm api

**Comment by adisuissa** [comment]
@tyxia can you please take a first-pass at the code changes?

**Comment by tyxia** [comment]
/assign @tyxia 

Sure. I will take a look

**Comment by StarryVae** [comment]
@tyxia @wbpcode friendly ping.

**Comment by wbpcode** [comment]
sorry for the delay. I will take a look tomorrow.

File: test/common/router/router_ratelimit_test.cc
Comment (tyxia) [inline test/common/router/router_ratelimit_test.cc:1094]: IMO, `QueryParameterValueMatchNotPresent` and `QueryParameterValueMatchPresent` are confusing, especially word `present`.

Basically,  `QueryParameterValueMatchNotPresent` is for case that `expect_not_match` and query parameters are indeed not matched (i.e., not_same_value != test_value)
`QueryParameterValueMatchPresent` is for case that `expect_not_match` but query parameters are matched.

I would suggest change name to `QueryParameterValueMatchExpectNoMatch` and `QueryParameterValueMatchExpectNoMatchFailed`. 
WDYT?

**Review by tyxia — CHANGES_REQUESTED** [review]
Thanks. Nice work! Mostly LGTM, just some nits.

Forward thinking, on the high level, [CEL expression](https://github.com/kyessenov/envoy/blob/a5389867fe536ab9676e61fe243c917f394934b7/source/extensions/filters/common/expr/context.cc#L109) could potentially be a good candidate to unify the matching methods here like `headerValue` and `QueryParameterValue`. But I think that is totally out of scope of this change and this PR matches existing styles/behaviors in rateLimit.

File: source/common/router/router_ratelimit.h
Comment (StarryVae) [reply source/common/router/router_ratelimit.h]: thanks, done

File: test/common/router/router_ratelimit_test.cc
Comment (StarryVae) [reply test/common/router/router_ratelimit_test.cc]: thanks, done

**Review by tyxia — DISMISSED** [review]
Thanks!

File: source/common/router/router_ratelimit.cc
Comment (wbpcode) [inline source/common/router/router_ratelimit.cc:255]: nit: should we use `parseAndDecodeQueryString` here to decode escaped parameter value?

**Review by wbpcode — COMMENTED** [review]
LGTM overall. Only one nit comment. 

/wait-any

File: source/common/router/router_ratelimit.cc
Comment (StarryVae) [reply source/common/router/router_ratelimit.cc:255]: yes, i think `parseAndDecodeQueryString` is more appropriate, thanks.

**Comment by StarryVae** [comment]
@adisuissa @wbpcode PTAL again, thanks.

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks.

**Comment by wbpcode** [comment]
ping cc @adisuissa for API tag.

**Review by adisuissa — COMMENTED** [review]
/lgtm api

**Comment by StarryVae** [comment]
ci is passed cc @wbpcode

**Pull request merged**
